Solving guidance page

bertiewooster opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Add a solving guidance page applicable to many types of solving.

#### Other comments
Pulled in content previously in #23638 and #23768

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* other
  * A new page on guidance applicable to many types of solving was added to the Guides section.

<!-- END RELEASE NOTES -->

bertiewooster committed:
Initial setup of solving guidance page

bertiewooster committed:
Title page

bertiewooster committed:
Move in "Use exact values" section from
solving system of inequalities branch

bertiewooster committed:
Import parse a string representing the equation

bertiewooster committed:
Impport "Equations with no analytical solution"

bertiewooster committed:
Add ref .. _symbolic-computation:

bertiewooster committed:
Link from solving main page to
solving guidance page

bertiewooster committed:
Remove blank, inapplicable sections

bertiewooster committed:
Set guides/index toc maxdepth to 1

bertiewooster committed:
Move solving guidance link before
table of solving types

bertiewooster committed:
Remove H2, promote three H3 to H2

asmeurer commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,141 @@
+>>> from sympy import symbols, solve, pi, Rational
+>>> x = symbols('x')
+>>> one_third = Rational(1, 3)
+>>> solve((x >= one_third, x**2 <= pi), x)
```
Let's just use a regular equation for this.

oscarbenjamin commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,141 @@
+({class}`CRootOf <sympy.polys.rootoftools.ComplexRootOf>` represents an 
+indexed complex root of a polynomial.)
+
+## Use exact values
```
Do you mean something like:
```
solve(x**sqrt(2).n() - 2)
```

oscarbenjamin commented on doc/src/guides/solving/solving-guidance.md:
````diff
@@ -0,0 +1,141 @@
+0.739085133215161
+```
+
+If you receive non-closed-form solutions, you can evaluate them numerically 
+using {func}`evalf n <sympy.core.evalf.EvalfMixin.n>`:
+
+```py
+>>> from sympy import solve
+>>> from sympy.abc import x
+>>> solutions = solve(x**5 - x - 1, x, dict=True)
+>>> print(solutions)
+[{x: CRootOf(x**5 - x - 1, 0)}, {x: CRootOf(x**5 - x - 1, 1)}, {x: CRootOf(x**5 - x - 1, 2)}, {x: CRootOf(x**5 - x - 1, 3)}, {x: CRootOf(x**5 - x - 1, 4)}]
+>>> [solution[x].n(3) for solution in solutions]
+[1.17, -0.765 - 0.352*I, -0.765 + 0.352*I, 0.181 - 1.08*I, 0.181 + 1.08*I]
````
Rather than "non-closed form" it would be better to say "RootOf" or something more specific.

oscarbenjamin replied:
It can also be useful to use `evalf` even if the result is closed form:
```python
solve(x**4 + 10*x**2 + x + 1, x)
```
The `solve` function will generally try to return exact solutions if its inputs are exact. Sometimes exact numeric solutions can have very complicated representations though but as long as they are numeric (do not involve symbols) then evalf should be able to reduce them to a simple albeit approximate form.

Of course if you did know that you had a polynomial and therefore that your solution was going to come out in terms of RootOf then solve is not the right function anyway: `nroots`, `real_roots`, `Poly.all_roots`, `roots`. We have many options for solving a univariate polynomial with different strengths and weaknesses.

RootOf specifically as well as the Abel-Ruffini theorem etc is something that deserves special mention in the section on polynomials but could be linked to from other places because the same properties apply general nonlinear equation solvers like solve and nonlinsolve whenever polynomials arise in the equations.

bertiewooster committed:
Use an equation for exact example
Mention CRootOf in narrative

bertiewooster committed:
Give example where solve fails
given numerical, inexact input

bertiewooster committed:
Add note about numeric solutions
Give numeric value for sqrt(2)

bertiewooster committed:
Add example where solve solution is cumbersome
so evalf may be preferable
Add links for some SymPy entities

bertiewooster committed:
Merge branch 'master' into solving-guidance

bertiewooster committed:
Merge branch 'master' into solving-guidance

bertiewooster committed:
Move content of Options that can speed up solve
from solve-equation-algebraically to solving-guidance

bertiewooster committed:
Use dict=True for all solve()

bertiewooster committed:
Use auto-generated heading slug for
Options that can speed up solve
per format at
https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#anchor-slug-structure

bertiewooster committed:
Fix doctest for dict=True

bertiewooster commented:
### Tangential question about MyST auto-generated targets
Instead of adding a custom target to a documentation file, I discovered that [MyST auto-creates targets based on headings](https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#auto-generated-header-anchors) for pages written in Markdown. I also learned that we can set the level down to which MyST will create targets using the [`conf.py` option myst_heading_anchors = 2](https://github.com/sympy/sympy/blob/master/doc/src/conf.py#L116), so we're currently set to go down to `h2` aka `##` in Markdown. It wasn't necessary for this case, but I thought we might sometimes want to go down to level 3. Is there any reason not to set `myst_heading_anchors = 3`, either now (preemptively) or if we need it for a specific target in the future? (I'm guessing it might take a bit longer to build the docs; not sure it would be noticeable).

bertiewooster committed:
Use Title Case

bertiewooster commented:
This [solving guidance page](https://output.circle-artifacts.com/output/job/d610c21e-fcc9-4f9f-9fc6-171148d2204e/artifacts/0/doc/_build/html/guides/solving/solving-guidance.html) is ready for final review, or merging if there are no more comments. Other solving pull requests (additional pages such as #23802) will link to this page, so I'd appreciate folks prioritizing this amongst the solving page pull requests.

bertiewooster committed:
Add note and link to Python numbers vs. SymPy Numbers

bertiewooster committed:
Clarify links for Use Exact Values

asmeurer commented:
> Tangential question about MyST auto-generated targets
Instead of adding a custom target to a documentation file, I discovered that [MyST auto-creates targets based on headings](https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#auto-generated-header-anchors) for pages written in Markdown. I also learned that we can set the level down to which MyST will create targets using the [conf.py option myst_heading_anchors = 2](https://github.com/sympy/sympy/blob/master/doc/src/conf.py?rgh-link-date=2022-08-07T19%3A55%3A31Z#L116), so we're currently set to go down to h2 aka ## in Markdown. It wasn't necessary for this case, but I thought we might sometimes want to go down to level 3. Is there any reason not to set myst_heading_anchors = 3, either now (preemptively) or if we need it for a specific target in the future? (I'm guessing it might take a bit longer to build the docs; not sure it would be noticeable).

Is the issue just performance? If so, we should try it and see if it makes a difference.

Or is the issue that it would more like lead to ambiguous targets if two pages have the same header names? How does it handle that case?

asmeurer commented on doc/src/guides/solving/solving-guidance.md:
````diff
@@ -0,0 +1,221 @@
+```py
+>>> from sympy import symbols, solve
+>>> x = symbols('x')
+>>> solutions = solve(x**4 + 10*x**2 + x + 1, x, dict=True); solutions
````
I'm not a fan of the semicolon thing in doctest examples. Just write `>>> solutions` on another line.

asmeurer commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,221 @@
+>>> from sympy import symbols, solve
+>>> x = symbols('x')
+>>> solutions = solve(x**4 + 10*x**2 + x + 1, x, dict=True); solutions
+[{x: -sqrt(-20/3 + 56/(9*(1307/432 + sqrt(434607)*I/144)**(1/3)) + 2*(1307/432 + sqrt(434607)*I/144)**(1/3))/2 - sqrt(-40/3 - 2*(1307/432 + sqrt(434607)*I/144)**(1/3) + 2/sqrt(-20/3 + 56/(9*(1307/432 + sqrt(434607)*I/144)**(1/3)) + 2*(1307/432 + sqrt(434607)*I/144)**(1/3)) - 56/(9*(1307/432 + sqrt(434607)*I/144)**(1/3)))/2}, {x: sqrt(-20/3 + 56/(9*(1307/432 + sqrt(434607)*I/144)**(1/3)) + 2*(1307 ...
```
It's too bad that the docs don't line wrap this. You don't really get the impression of how unwieldy these are.

bertiewooster replied:
Yes, I had the same thought when I showed you the page when we met. I added a description of the scale of the exact solution, and to scroll horizontally to see all the terms.

asmeurer replied:
We could probably change the CSS to make it softwrap, but I don't know what that would do to other code examples throughout the docs.

asmeurer commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,221 @@
+
+## Numeric Solutions
+
+### Numeric Solutions by Choice
```
This is a string title name. It's not clear what "by choice" is supposed to mean here.

asmeurer commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,221 @@
+The vast majority of arbitrary nonlinear equations are not analytically
+solvable. The classes of equations that are solvable are basically:
+1. Linear equations
+2. Polynomials, except where limited by the Abel-Ruffini theorem
```
Link to Wikipedia for Abel-Ruffini.

bertiewooster replied:
I was considering a link, and didn't add one because there wasn't an explanation of it in our docs; this is a good idea.

asmeurer commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,221 @@
+5. A few other special cases that can be solved with something like the
+{class}`Lambert W function <sympy.functions.elementary.exponential.LambertW>` 
+
+SymPy may reflect that your equation has no solutions that can be expressed
```
Just to be clear, this could mean it can't be solved in closed-form or it could just mean that it isn't implemented, as the error name suggests. SymPy doesn't have any functionality to guarantee that an equation's solutions cannot be written in closed-form (in general, that's not a very easy thing to prove).

asmeurer commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,221 @@
+2. Polynomials, except where limited by the Abel-Ruffini theorem
+3. Equations that can be solved by inverting some transcendental functions
+4. Problems that can be transformed into the cases above (e.g., by turning
+trigonometric functions into polynomials)
```
I'm curious what this means (I'm assuming this list came from @oscarbenjamin or @smichr). 

Wouldn't the most general case be "equations that can be decomposed via any of the above", e.g., a polynomial in sin(x) can be inverted because both a polynomial and sin(x) can be inverted (but that is also true for any f(g(x)) if both are solvable)?

smichr replied:
Yes. You could replace `x` in a Lambert expression with a polynomial `p = f(x)` and get solutions for `p` and then, for each of them get a solution for `x` by solving `f(x) = solution value`.

asmeurer commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,221 @@
+
+If you receive non-closed-form solutions such as {class}`CRootOf
+<sympy.polys.rootoftools.ComplexRootOf>`, you can evaluate them numerically
+using {func}`evalf n <sympy.core.evalf.EvalfMixin.n>`:
```
I think Oscar mentioned this, but evalf is also useful even for closed-form solutions (especially when they're large, but even when they're not large, sometimes you just want to know the decimal value of a number).

asmeurer replied:
Is this supposed to be `evalf/n`? IMO we should just mention one to avoid confusion (I personally prefer `evalf`).

asmeurer commented on doc/src/guides/solving/solving-guidance.md:
````diff
@@ -0,0 +1,221 @@
+```
+
+({class}`CRootOf <sympy.polys.rootoftools.ComplexRootOf>` represents an indexed
+complex root of a polynomial.)
````
Maybe mention that when solve returns CRootOf, it typically means that SymPy could not express the solution in radicals (which usually means there isn't a solution in radicals, but not necessarily since SymPy's polynomial solvers aren't complete in this sense).

asmeurer commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,221 @@
+
+### Do Not Simplify Solutions by Using `simplify=False`
+
+Normally, {func}`~.solve` simplifies all but polynomials of order 3 or greater
```
Unless I'm mistaken, `simplify` is about simplifying the solutions, not the equations. I guess this is copied from the `solve` docstring. Maybe @smichr can confirm either way. The "polynomials of order 3" thing is an unnecessary detail to include here IMO.

asmeurer replied:
Oh I didn't read the rest of the sentence. So I guess it does both. Either way, I would just omit the technical details here, which can be found in the solve docstring. Just mention that simplify=False makes it not try to simplify the equation and solutions, which can sometimes slow things down if they are complicated.

asmeurer commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,221 @@
+## Parse a String Representing the Equation
+
+If you are creating the expression yourself, we [recommend against using parsing
+a string](
```
"against using string parsing to create expressions"

asmeurer commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,221 @@
+- create symbols from your variables
+-  use more mathematical (rather than standard Python) notation, for example the
+exponent operator can be parsed from `^` rather than having to use Python's
+`**`.
```
I guess as long as we're mentioning string parsing, we can mention quickly that SymPy supports [parsing LaTeX](https://docs.sympy.org/dev/modules/parsing.html#sympy.parsing.latex.parse_latex). In my experience, a lot of people find that feature useful. Perhaps it should also be mentioned in the "see also" for `parse_expr`.

bertiewooster committed:
Changes per review comments

smichr commented on doc/src/guides/solving/index.md:
```diff
@@ -5,6 +5,9 @@ The Python package SymPy can symbolically solve equations, differential equation
 linear equations, nonlinear equations, matrix problems, inequalities,
 Diophantine equations, and evaluate integrals. SymPy can also solve numerically.
 
+You can refer to [solving guidance](solving-guidance.md) applicable to many
+types of solving.
```
```suggestion
Refer to the document [here](solving-guidance.md) for guidance applicable to many
types of solving tasks.
```

smichr commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -56,11 +55,10 @@ Here are recommendations on when to use:
-SymPy will solve for the desired symbol. For example, you may expect the
-following to solve for $x$,
-and SymPy will solve for $y$:
+for either function. While this is optional for equations with a single symbol,
```
```suggestion
for either function. While this is optional for univariate equations,
```

smichr commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -56,11 +55,10 @@ Here are recommendations on when to use:
-and SymPy will solve for $y$:
+for either function. While this is optional for equations with a single symbol,
+it is a good practice because it ensures SymPy will solve for the desired
+symbol. For example, you may expect the following to solve for $x$, and SymPy
+will solve for $y$:
```
```suggestion
symbol. For example, you might be interested in a solution for $x$, but SymPy
solves for $y$:
```

smichr commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -80,26 +78,24 @@ Specifying the variable to solve for ensures that SymPy solves for it:
-unless you use `dict=True` to ensure the result will be formatted as a
-dictionary. We recommend using `dict=True`, especially if you want to
-extract information from the result programmatically.
+{func}`~.solve` produces various output formats depending on the answer, unless
+you use `dict=True` to ensure the result will be formatted as a dictionary. We
+recommend using `dict=True`, especially if you want to extract information from
+the result programmatically.
```
```suggestion
{func}`~.solve` produces a variety of output as explained in :ref:`solve_output`. Using
`dict=True` will give a consistent output format which is especially important when
extracting information about the solution programmatically.
```

smichr commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -112,10 +108,9 @@ $expression = 0$.
-or which you think of as an equality.
+Put your equation into `Eq` form, then solve the `Eq`. This approach is
+convenient if you are interactively solving an equation which you already have
+in the form of an equation, or which you think of as an equality.
```
```suggestion
in the form of an equation, or which you think of as an equality. It also
help to prevent sign errors when subtracting one side from the other.
```

smichr commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -235,25 +230,24 @@ However, {func}`~.solve` will return only a finite number of solutions:
-equation, here $2\pi$.
+{func}`~.solve` tries to return just enough solutions so that all (infinitely
+many) solutions can generated from the returned solutions by adding integer
+multiples of the {func}`~.periodicity` of the equation, here $2\pi$.
```
nice reference there to `periodicity`

smichr commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,245 @@
+
+## Parse a String Representing the Equation
+
+If you are creating the expression yourself, we [against using string parsing to
```
```suggestion
If you are creating the expression yourself, we advise [against using string parsing to
```

smichr commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,245 @@
+
+You can parse a string representing the equation into a form that SymPy can
+understand (for example, {any}`Eq() <sympy.core.relational.Eq>` form), then
+apply solve the parsed expression. Parsing an equation from a string requires
```
```suggestion
solve the parsed expression. Parsing an equation from a string requires
```

smichr commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,245 @@
+>>> print(parsed)
+Eq(x**2, y)
+>>> solutions = solve(parsed, x, dict=True)
+>>> print(solutions)
+[{x: -sqrt(y)}, {x: sqrt(y)}]
+>>> for solution in solutions:
+...     for key, val in solution.items():
+...         print(val)
+-sqrt(y)
+sqrt(y)
```
```suggestion
>>> [s[x] for s in solutions]
[-sqrt(y), sqrt(y)]
```

smichr replied:
This seems to be a little out of place. Perhaps this should go up right after the suggestion about using `dict=True`.

bertiewooster replied:
Thanks. This was a result of rearranging content. I realized the `dict=True` guidance was in another page ([Solve An Equation Algebraically](https://docs.sympy.org/dev/guides/solving/solve-equation-algebraically.html)) so I moved that to Solving Guidance and then was able to follow your suggestion.

smichr commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,245 @@
+>>> parsed = parse_expr(expr, transformations="all")
+>>> print(parsed)
+Eq(x**2, y)
+>>> solutions = solve(parsed, x, dict=True)
+>>> print(solutions)
+[{x: -sqrt(y)}, {x: sqrt(y)}]
+>>> for solution in solutions:
+...     for key, val in solution.items():
+...         print(val)
+-sqrt(y)
+sqrt(y)
+>>> solutions_set = solveset(parsed, x)
+>>> print(solutions_set)
+{-sqrt(y), sqrt(y)}
```
```suggestion
```

smichr replied:
Since you are only showing how to parse strings, I don't think it's necessary to repeat the solution process.

smichr commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,245 @@
+trigonometric functions into polynomials)
+5. A few other special cases that can be solved with something like the
+{class}`Lambert W function <sympy.functions.elementary.exponential.LambertW>`
+6. Equations that can be decomposed via any of the above
```
Perhaps we could get a link to `decompose` here.

smichr commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,245 @@
+The vast majority of arbitrary nonlinear equations are not analytically
+solvable. The classes of equations that are solvable are basically:
+1. Linear equations
+2. Polynomials, except where limited by the [Abel-Ruffini
```
Perhaps a mention to Groebner basis reduction for systems of polynomial equations? @oscarbenjamin is always dazzling me with demonstrations of this on SO. I wish I understood more but a link to that function might help people be aware of this technique.

asmeurer commented on doc/src/guides/solving/solving-guidance.md:
```diff
@@ -0,0 +1,245 @@
+
+In other situations, even if the exact solution has few terms, you may want a
+numeric solution so you know its approximate numerical value. For example, it
+may be difficult to estimate that $\sqrt{2} * e^{\pi}/2$ is approximately $16$:
```
```suggestion
may be difficult to estimate that $\sqrt{2} e^{\pi}/2$ is approximately $16$:
```

bertiewooster committed:
Update doc/src/guides/solving/index.md

Co-authored-by: Christopher Smith <smichr@gmail.com>

bertiewooster committed:
Update doc/src/guides/solving/solve-equation-algebraically.md

Co-authored-by: Christopher Smith <smichr@gmail.com>

bertiewooster committed:
Update doc/src/guides/solving/solve-equation-algebraically.md

Co-authored-by: Christopher Smith <smichr@gmail.com>

bertiewooster committed:
Update doc/src/guides/solving/solve-equation-algebraically.md

Co-authored-by: Christopher Smith <smichr@gmail.com>

bertiewooster committed:
Update doc/src/guides/solving/solving-guidance.md

Co-authored-by: Christopher Smith <smichr@gmail.com>

bertiewooster committed:
Update doc/src/guides/solving/solving-guidance.md

Co-authored-by: Christopher Smith <smichr@gmail.com>

bertiewooster committed:
Update doc/src/guides/solving/solving-guidance.md

Co-authored-by: Aaron Meurer <asmeurer@gmail.com>

bertiewooster committed:
Incorporate review comments

bertiewooster committed:
Update doc/src/guides/solving/solving-guidance.md

Co-authored-by: Christopher Smith <smichr@gmail.com>

bertiewooster committed:
Update doc/src/guides/solving/solving-guidance.md

Co-authored-by: Christopher Smith <smichr@gmail.com>

bertiewooster committed:
Move some solving guidance from solve-equation-algebraically to solving-guidance

bertiewooster committed:
Capitalize headings in Solve an Equation Algebraically

bertiewooster committed:
Merge branch 'master' into solving-guidance
Merge master into solving-guidance to get doc/src/explanation/solve_output.rst

bertiewooster committed:
Improve some reference links

bertiewooster committed:
Change to "no closed-form solution" with transcendental example
"analytical" -> "closed-form"

bertiewooster commented:
> > Tangential question about MyST auto-generated targets
> > Instead of adding a custom target to a documentation file, I discovered that [MyST auto-creates targets based on headings](https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#auto-generated-header-anchors) for pages written in Markdown. I also learned that we can set the level down to which MyST will create targets using the [conf.py option myst_heading_anchors = 2](https://github.com/sympy/sympy/blob/master/doc/src/conf.py?rgh-link-date=2022-08-07T19%3A55%3A31Z#L116), so we're currently set to go down to h2 aka ## in Markdown. It wasn't necessary for this case, but I thought we might sometimes want to go down to level 3. Is there any reason not to set myst_heading_anchors = 3, either now (preemptively) or if we need it for a specific target in the future? (I'm guessing it might take a bit longer to build the docs; not sure it would be noticeable).
> 
> Is the issue just performance? If so, we should try it and see if it makes a difference.

I built the docs from scratch locally at various levels of myst_heading_anchors and the results were the same:
myst_heading_anchors = 2: 45 seconds
myst_heading_anchors = 3: 45 seconds
myst_heading_anchors = 6: 45 seconds

> Or is the issue that it would more like lead to ambiguous targets if two pages have the same header names? How does it handle that case?

I should have used more precise language: [MyST auto-generates anchors](https://myst-parser.readthedocs.io/en/latest/syntax/optional.html?highlight=slug#auto-generated-header-anchors) rather than targets. That means you have to specify the page that contains them for example `[link text](./syntax.md#core-syntax)`, so there are no conflicts if the same header appears on multiple pages.

I asked if there were any other downsides on Write the Docs Slack (technical documentation forum) and the Executable Books Project lead developer responded "should be all good".

So I'll plan to increase `myst_heading_anchors` in one of these pull requests. Thinking I'll just set it to `6` to avoid any future possibility of a heading not having an anchor auto-generated for it. While I doubt we'd ever go six levels deep, seems like any downsides are, at worst, negligible.

smichr commented:
I sure hope others find these useful. I appreciate your work on these.
